pkg/odo/cli/component/create.go

```diff
@@ -881,6 +881,101 @@ func (co *CreateOptions) Validate() (err error) {
 	return nil
 }
 
+func downloadGitProject(starterProject *devfilev1.StarterProject, starterToken, path string) error {
+
+	var projectSource devfilev1.GitLikeProjectSource
+	if starterProject.Git != nil {
+		projectSource = starterProject.Git.GitLikeProjectSource
+	} else {
+		projectSource = starterProject.Github.GitLikeProjectSource
+	}
+
+	remoteName, remoteUrl, revision, err := parsercommon.GetDefaultSource(projectSource)
+	if err != nil {
+		return errors.Wrapf(err, "unable to get default project source for starter project %s", starterProject.Name)
+	}
+
+	refName := plumbing.ReferenceName(revision)
+
+	if plumbing.IsHash(revision) {
+		log.Warning("Specifying commit in 'revision' is not yet supported in odo.")
+		// overriding revision to empty as we do not support this
+		revision = ""
+	}
+
+	if revision != "" {
+		// lets consider revision to be a branch name first
+		refName = plumbing.NewBranchReferenceName(revision)
+	}
+
+	downloadSpinner := log.Spinnerf("Downloading starter project %s from %s", starterProject.Name, remoteUrl)
+	defer downloadSpinner.End(false)
+
+	cloneOptions := &git.CloneOptions{
+		URL:           remoteUrl,
+		RemoteName:    remoteName,
+		ReferenceName: refName,
+		SingleBranch:  true,
+		// we don't need history for starter projects
+		Depth: 1,
+	}
+
+	if starterToken != "" {
+		cloneOptions.Auth = &http.BasicAuth{
+			Username: registryUtil.RegistryUser,
+			Password: starterToken,
+		}
+	}
+
+	originalPath := ""
+	if starterProject.SubDir != "" {
+		originalPath = path
+		path, err = ioutil.TempDir("", "")
+		if err != nil {
+			return err
+		}
+	}
+
+	_, err = git.PlainClone(path, false, cloneOptions)
+
+	if err != nil {
+
+		// it returns the following error if no matching ref found
+		// if we get this error, we are trying again considering revision as tag, only if revision is specified.
+		if _, ok := err.(git.NoMatchingRefSpecError); !ok || revision == "" {
+			return err
+		}
+
+		// try again to consider revision as tag name
+		cloneOptions.ReferenceName = plumbing.NewTagReferenceName(revision)
+		// remove if any .git folder downloaded in above try
+		_ = os.RemoveAll(filepath.Join(path, ".git"))
+		_, err = git.PlainClone(path, false, cloneOptions)
```

Comment:
Probably silly/dumb question - Have we tried cloning using a tag instead of branch name? I'm asking because I see no change in values between this call and the same call a few lines above.

Reply:
replied in the comment above